File: c/tree/house-robber-iii.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int rob(struct TreeNode* root) {
    if(root == NULL)
        return 0;

    int m1 = rob(root->left) + rob(root->right);
    int m2 = root->val;
    if(root->left)
        m2 += rob(root->left->left) + rob(root->left->right);
    if(root->right)
        m2 += rob(root->right->left) + rob(root->right->right);

    return m2 > m1 ? m2 : m1;
}

File: python/tree/house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def dfs(self, root):
        if root is None:
            return 0,0

        left,not_left = self.dfs(root.left)
        right, not_right = self.dfs(root.right)

        cur = not_left + not_right + root.val
        not_cur = max(left,not_left) + max(right,not_right)

        return cur, not_cur

    def rob(self, root):
        cur,not_cur = self.dfs(root)

        return max(cur,not_cur)
